dali/ft/filecopy.cpp

```diff
@@ -3979,63 +4012,82 @@ bool FileSprayer::calcUsePull() const
     {
-        LOG(MCdebugInfo, "Use push since explicitly specified");
-        return false;
-    }
+        // In BM the sources or targets may not be HPCC environment machines and/or might not be able to run ftslave
+        ForEachItemIn(idx2, sources)
+        {
+            if (!sources.item(idx2).canPush())
+            {
+                StringBuffer s;
+                sources.item(idx2).filename.queryIP().getHostText(s);
+                LOG(MCdebugInfo, "Use pull operation because %s cannot push", s.str());
+                return true;
+            }
+        }
 
-    ForEachItemIn(idx2, sources)
-    {
-        if (!sources.item(idx2).canPush())
+        if (!canLocateSlaveForNode(sources.item(0).filename.queryIP()))
         {
             StringBuffer s;
-            sources.item(idx2).filename.queryIP().getHostText(s);
-            LOG(MCdebugInfo, "Use pull operation because %s cannot push", s.str());
+            sources.item(0).filename.queryIP().getHostText(s);
+            LOG(MCdebugInfo, "Use pull operation because %s doesn't appear to have an ftslave", s.str());
             return true;
         }
-    }
-    if (!canLocateSlaveForNode(sources.item(0).filename.queryIP()))
-    {
-        StringBuffer s;
-        sources.item(0).filename.queryIP().getHostText(s);
-        LOG(MCdebugInfo, "Use pull operation because %s doesn't appear to have an ftslave", s.str());
-        return true;
-    }
 
-    ForEachItemIn(idx, targets)
-    {
-        if (!targets.item(idx).canPull())
+        ForEachItemIn(idx, targets)
+        {
+            if (!targets.item(idx).canPull())
+            {
+                StringBuffer s;
+                targets.item(idx).queryIP().getHostText(s);
+                LOG(MCdebugInfo, "Use push operation because %s cannot pull", s.str());
+                return false;
+            }
+        }
+
+        if (!canLocateSlaveForNode(targets.item(0).queryIP()))
         {
             StringBuffer s;
-            targets.item(idx).queryIP().getHostText(s);
-            LOG(MCdebugInfo, "Use push operation because %s cannot pull", s.str());
+            targets.item(0).queryIP().getHostText(s);
+            LOG(MCdebugInfo, "Use push operation because %s doesn't appear to have an ftslave", s.str());
             return false;
         }
     }
 
-    if (!canLocateSlaveForNode(targets.item(0).queryIP()))
+    bool pullRequested = options->hasProp(ANpull) ? options->getPropBool(ANpull) : false;
+    bool pushRequested = options->hasProp(ANpush) ? options->getPropBool(ANpush) : false;
+    if (!targetSupportsConcurrentWrite)
     {
-        StringBuffer s;
-        targets.item(0).queryIP().getHostText(s);
-        LOG(MCdebugInfo, "Use push operation because %s doesn't appear to have an ftslave", s.str());
-        return false;
+        // always pull
+        if (pushRequested)
+            IWARNLOG("Ignoring push option as targets < sources and target does not support concurrent write");
```

Comment:
I think this could occur if targets > sources - including sources==1 where push may make sense, especially if it is small.

Reply:
Multiple writes could be involved even if targets < sources (see comment on line 4072)

Reply:
if >1 source , and pushing, then each source pusher may write to same target, e.g. 2 source pushing to 3 targets, each source will push to the middle target ?

sources == 1 .. ok that is special and will only ever be pushing 1 partitions to single target ?

so special case source==1 here and leave with push preference if specified, otherwise, calc size and push if < threshold and pull otherwise?